src/sentry/auth/providers/saml2.py

```diff
@@ -349,196 +146,187 @@ def dispatch(self, request, organization_slug):
-        # return None   # TODO  If I return None, then a loop after execute the config
-        # happens from organizations/<org>/auth/ to /auth/login/  /<org>/
-        identity = {}
-        if state and 'contact' in state:
-            identity['id'] = state['contact']
-            identity['email'] = state['contact']
-        return identity
-
-    def build_saml_config(self, org_slug):
-        metadata_url = absolute_uri(
-            reverse('sentry-auth-organization-saml-metadata', args=[org_slug])
-        )
-        acs_url = absolute_uri(reverse('sentry-auth-organization-saml-acs', args=[org_slug]))
-        sls_url = absolute_uri(reverse('sentry-auth-organization-saml-sls', args=[org_slug]))
+        This value will be merged into the configuration by self.build_config,
+        however, should a attribute_mapping exist in the helper state at
+        configuration build time, these may be overriden.
+        """
+        return {}
 
-        saml_config = {}
-        saml_config['strict'] = True
-        saml_config['idp'] = self.extract_parsed_data_from_idp_data(self.config)
-        saml_config['sp'] = {
-            "entityId": metadata_url,
-            "assertionConsumerService": {
-                "url": acs_url,
-                "binding": OneLogin_Saml2_Constants.BINDING_HTTP_POST
-            },
-            "singleLogoutService": {
-                "url": sls_url,
-                "binding": OneLogin_Saml2_Constants.BINDING_HTTP_REDIRECT
-            }
-        }
+    def build_config(self, state):
+        config = state
 
-        saml_config['security'] = self.extract_parsed_data_from_advanced_data(self.config)
+        # Default attriute mapping if none bound
+        if 'attribute_mapping' not in config:
+            config['attribute_mapping'] = self.attribute_mapping()
 
-        sp_entity_id = saml_config['security'].get('spEntityId', None)
-        if sp_entity_id:
-            saml_config['sp']['entityId'] = sp_entity_id
-            del saml_config['security']['spEntityId']
+        return config
 
-        sp_x509cert = saml_config['security'].get('spx509cert', None)
-        if sp_x509cert:
-            saml_config['sp']['x509cert'] = sp_x509cert
-            del saml_config['security']['spx509cert']
+    def build_identity(self, state):
+        raw_attributes = state['auth_attributes']
+        attributes = {}
 
-        sp_private_key = saml_config['security'].get('spPrivateKey', None)
-        if sp_private_key:
-            saml_config['sp']['privateKey'] = sp_private_key
-            del saml_config['security']['spPrivateKey']
+        # map configured provider attributes
+        for key, provider_key in iteritems(self.config['attribute_mapping']):
+            attributes[key] = raw_attributes[provider_key][0]
 
-        return saml_config
+        name = (attributes[k] for k in (Attributes.FIRST_NAME, Attributes.LAST_NAME))
+        name = ' '.join(filter(lambda n: n is not None, name))
```

Comment:
If you use `filter(None, name)` this would be more correct imo, since this will just filter out all falsey values rather than strictly Nones. Like, if last name were an empty string value.

```python
>>> filter(None, [None, 1, 0, ''])
[1]
```

Reply:
perfect